### PR TITLE
Texture Filtering v2

### PR DESCRIPTION
**Pull request by BreadFish64**
Follow-up to #5017

I was having some trouble getting texture filtering and format reinterpretation working together. I decided I wasn't happy with the current texture filtering design, so I did a complete self-review and made some major changes. Hopefully, this should also fix some of the reported bugs with texture filtering that I haven't reproduced

For more info check the commit messages.

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/5166)

### CODE REVIEW

File: src/video_core/renderer_opengl/gl_rasterizer_cache.h [hunk 1]
```diff
@@ -344,7 +133,11 @@ struct SurfaceWatcher {
     bool valid = false;
 };
 
+class RasterizerCacheOpenGL;
```
Comment (lioncash) [inline src/video_core/renderer_opengl/gl_rasterizer_cache.h:136]: Forward declarations should be situated near the top of the file. Ditto for other occurrences.

File: src/video_core/renderer_opengl/texture_filters/texture_filterer.cpp [hunk 2]
```diff
@@ -0,0 +1,79 @@
+/// Copyright 2020 Citra Emulator Project
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+
+#include <functional>
```
Comment (lioncash) [inline src/video_core/renderer_opengl/texture_filters/texture_filterer.cpp:5]: `<algorithm>` needs to be included.

File: src/video_core/renderer_opengl/texture_filters/texture_filterer.h [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+#include "video_core/renderer_opengl/texture_filters/texture_filter_base.h"
+
+namespace OpenGL {
+
+class TextureFilterer {
+public:
+    static constexpr std::string_view NONE = "none";
+
+    explicit TextureFilterer(std::string_view filter_name, u16 scale_factor);
+    // returns true if the filter actually changed
```
Comment (lioncash) [inline src/video_core/renderer_opengl/texture_filters/texture_filterer.h:23]: If a destructor is defaulted within the cpp file,

```cpp
#include "video_core/renderer_opengl/texture_filters/texture_filter_base.h"
```

can be replaced with a forward declaration of `TextureFilterBase`.

File: src/video_core/renderer_opengl/texture_filters/texture_filterer.cpp [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+#include "video_core/renderer_opengl/texture_filters/bicubic/bicubic.h"
+#include "video_core/renderer_opengl/texture_filters/texture_filterer.h"
+#include "video_core/renderer_opengl/texture_filters/xbrz/xbrz_freescale.h"
+
+namespace OpenGL {
+
+using TextureFilterContructor = std::function<std::unique_ptr<TextureFilterBase>(u16)>;
+
+template <typename T>
+std::pair<std::string_view, TextureFilterContructor> FilterMapPair() {
```
Comment (lioncash) [inline src/video_core/renderer_opengl/texture_filters/texture_filterer.cpp:17]:
```suggestion
static std::pair<std::string_view, TextureFilterContructor> FilterMapPair() {
```

or enclose all internal details with an anonymous namespace to give this internal linkage.

File: src/video_core/renderer_opengl/texture_filters/texture_filterer.h (hunk 3 above)
Comment (BreadFish64) [reply src/video_core/renderer_opengl/texture_filters/texture_filterer.h:23]: I tried this, but unfortunately the compiler errors with a message about deleting an incomplete type.

**Comment by BreadFish64** [comment]
I'm merging this quickly in the interest of unblocking texture format reinterpretation.

**Pull request merged**